### PR TITLE
Added posibility of calculate record status time wtihout sla record

### PR DESCRIPTION
**Pull request by konadrian**
image

### CODE REVIEW

File: app/Utils/ServiceContracts.php [hunk 1]
```diff
@@ -380,6 +436,33 @@ public static function getDiffFromServiceContracts(string $start, string $end, i
...
+				return round(\App\Fields\DateTime::getDiff($start, $end, 'minutes'));
+			}
+			$time = 0;
+			foreach ($rules as $row) {
+				$time += self::businessTime($start, $end, $row['working_hours_from'], $row['working_hours_to'], explode(',', $row['working_days']), !empty($row['holidays']));
+			}
+			return $time;
+		}
+		if (!($diff = self::getDiffFromDefaultBusinessHours($start, $end))) {
+			$diff = round(\App\Fields\DateTime::getDiff($start, $end, 'minutes'));
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]: ?

File: app/Utils/ServiceContracts.php [hunk 2]
```diff
@@ -380,6 +436,33 @@ public static function getDiffFromServiceContracts(string $start, string $end, i
...
+	 * @param string               $end
+	 * @param \Vtiger_Record_Model $recordModel Record the model that will be updated
+	 *
+	 * @return int
+	 */
+	public static function getDiffFromSlaPolicy(string $start, string $end, \Vtiger_Record_Model $recordModel): int
+	{
+		if ($rules = self::getSlaPolicyRulesForRecord($recordModel)) {
+			if (isset($rules['id'])) {
+				return round(\App\Fields\DateTime::getDiff($start, $end, 'minutes'));
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]: ?

File: app/Utils/ServiceContracts.php [hunk 3]
```diff
@@ -380,6 +436,33 @@ public static function getDiffFromServiceContracts(string $start, string $end, i
...
+	 * @param string               $start
+	 * @param string               $end
+	 * @param \Vtiger_Record_Model $recordModel Record the model that will be updated
+	 *
+	 * @return int
+	 */
+	public static function getDiffFromSlaPolicy(string $start, string $end, \Vtiger_Record_Model $recordModel): int
+	{
+		if ($rules = self::getSlaPolicyRulesForRecord($recordModel)) {
+			if (isset($rules['id'])) {
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]: ?

File: app/Utils/ServiceContracts.php [hunk 4]
```diff
@@ -265,6 +265,59 @@ public static function getRulesForServiceContracts(int $serviceContractId, \Vtig
...
+	}
+
+	/**
+	 * Get sla policy by crm id.
+	 *
+	 * @param int $sourceModuleId
+	 *
+	 * @return array
+	 */
+	public static function getSlaPolicyForRecord(int $sourceModuleId): array
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]:
```suggestion
	public static function getSlaPolicyForModule(int $moduleId): array
```

File: app/Utils/ServiceContracts.php [hunk 5]
```diff
@@ -265,6 +265,59 @@ public static function getRulesForServiceContracts(int $serviceContractId, \Vtig
...
+	{
+		if (\App\Cache::has('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId)) {
+			$rows = \App\Cache::get('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId);
+		} else {
+			$rows = (new \App\Db\Query())->from('s_#__sla_policy')->where(['tabid' => $sourceModuleId, 'available_for_record_time_count' => 1])->all(\App\Db::getInstance('admin'));
+			\App\Cache::save('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId, $rows);
+		}
+		return (new \App\Db\Query())->from('s_#__sla_policy')->where(['tabid' => $sourceModuleId, 'available_for_record_time_count' => 1])->all(\App\Db::getInstance('admin'));
```
Comment (mariuszkrzaczkowski) [inline app/Utils/ServiceContracts.php]:
```suggestion
		return $rows;
```

**Review by mariuszkrzaczkowski — CHANGES_REQUESTED** [review]

File: app/Utils/ServiceContracts.php [hunk 6]
```diff
@@ -272,10 +272,10 @@ public static function getRulesForServiceContracts(int $serviceContractId, \Vtig
 	 *
 	 * @return array
 	 */
-	public static function getSlaPolicyRulesForRecord(\Vtiger_Record_Model $recordModel): array
+	public static function getSlaPolicyRulesForForModule(\Vtiger_Record_Model $recordModel): array
```
Comment (rskrzypczak) [inline app/Utils/ServiceContracts.php]: getSlaPolicyRules**ForFor**Module ?

File: app/Utils/ServiceContracts.php [hunk 7]
```diff
@@ -303,19 +303,19 @@ public static function getSlaPolicyRulesForRecord(\Vtiger_Record_Model $recordMo
...
 	{
-		if (\App\Cache::has('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId)) {
-			$rows = \App\Cache::get('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId);
+		if (\App\Cache::has('UtilsServiceContracts::getSlaPolicyForModule', $moduleId)) {
+			$rows = \App\Cache::get('UtilsServiceContracts::getSlaPolicyForModule', $moduleId);
 		} else {
-			$rows = (new \App\Db\Query())->from('s_#__sla_policy')->where(['tabid' => $sourceModuleId, 'available_for_record_time_count' => 1])->all(\App\Db::getInstance('admin'));
-			\App\Cache::save('UtilsServiceContracts::getSlaPolicyForRecord', $sourceModuleId, $rows);
+			$rows = (new \App\Db\Query())->from('s_#__sla_policy')->where(['tabid' => $moduleId, 'available_for_record_time_count' => 1])->all(\App\Db::getInstance('admin'));
+			\App\Cache::save('UtilsServiceContracts::getSlaPolicyForModule', $moduleId, $rows);
```
Comment (rskrzypczak) [inline app/Utils/ServiceContracts.php:316]: where you clean this cache?

**Pull request merged**